cirq/qis/states.py

```diff
@@ -343,39 +368,36 @@ def infer_qid_shape(*states: 'cirq.QUANTUM_STATE_LIKE') -> Tuple[int, ...]:
 
     # if all states are specified as integers, the shape is ambiguous
     if not non_integer_states:
-        raise ValueError('The qid shape of the given states is ambiguous.')
+        raise ValueError(
+            'The qid shape of the given states is ambiguous, '
+            'since all states were specified as integers.'
+        )
 
     # attempt to infer qid shape from non-integer states
     potential_shapes = _potential_qid_shapes(non_integer_states[0])
     for state in non_integer_states[1:]:
         potential_shapes = potential_shapes.intersection_subset(_potential_qid_shapes(state))
-    if potential_shapes.is_ambiguous():
-        raise ValueError('The qid shape of the given states is ambiguous.')
     qid_shape = potential_shapes.infer_qid_shape()
     if qid_shape is None:
-        raise ValueError('Failed to infer the qid shape of the given states.')
+        raise ValueError(
+            'Failed to infer the qid shape of the given states. '
+            'Either there is no compatible qid shape for the given states, or there is one '
+            'but the heuristic failed to find it.'
+        )
```

Comment:
The rest of the new error messages look great. This one reads really funny, however. "Either there is none or there is one" is a bad excuse for a program to make :) :) If this is truly an edge case, I'd recommend admitting as much

> Failed to infer the qid shape of the given states. If you think there is an unambiguous result for this case, please raise a GitHub issue.

Otherwise, parcel out the conditionals that can lead here: e.g. 

> Could not determine the total (unfactorized) Hilbert space dimension and could not find any explicit Qid shapes.

Reply:
I took your first suggestion; I do believe that this case would be reached only in unusual circumstances.